src/fsharp/lib.fs

```diff
@@ -394,6 +394,10 @@ let inline cacheOptRef cache f =
        cache := Some res
        res 
 
+let inline tryGetCacheValue cache =
+    match box cache.cacheVal with
```

Comment:
Is this isNull?

Reply:
Almost, but `cacheVal` is not nullable.
Definitions:
```fsharp
type NonNullSlot<'T> = 'T
type cache<'T> = { mutable cacheVal: 'T NonNullSlot }
```
So we can't use `isNull` as it requires nullability.

First time I've seen the box technique though. It's also done with the function above:
```fsharp
let inline cached cache resf = 
    match box cache.cacheVal with 
    | null -> 
...
```
Another alternative is `obj.ReferenceEquals`.

Reply:
Hmm. Taking a look at it again. The abbreviation is just for documentation.

If we got rid of the `box`, it would make the `cached` function require the typar `'T` to be nullable using a null constraint.

Reply:
it is isnull, the jit has a special optimization to make this fast.  @dsyme has the details.

Reply:
We do it quite a few places in the codebase

Reply:
> it is isnull, the jit has a special optimization to make this fast. @dsyme has the details.

Yes, though always worth having someone familiar with x86 assembly double check :)